src/odemis/acq/stitching/_tiledacq.py

```diff
@@ -926,3 +928,194 @@ def acquireTiledArea(streams, stage, area, overlap=0.2, settings_obs=None, log_p
+                              self.focus_rng[1] + self._focus.position.value["z"])
+
+        self.conf_level = 0.8
+        self.focusing_method = focusing_method
+        # The number of focus points are constant for now
+        # TODO: make it variable according to the size of the ROI, is it useful?
+        self.n_focus_points = (3, 3)
+        self._overlap = overlap
+        self._settings_obs = settings_obs
+        self._log_path = log_path
+        self._zlevels = zlevels
+        self._registrar = registrar
+        self._weaver = weaver
+
+    def cancel(self, future):
+        """
+        Canceler of acquisition task.
+        """
+        logging.debug("canceling acquisition overview...")
+
+        with future._task_lock:
+            if future._task_state == FINISHED:
+                return False
+            future._task_state = CANCELLED
+            future.running_subf.cancel()
+            logging.debug("acquisition overview cancelled.")
+        return True
+
+    def estimate_time(self, roi_idx=0, actual_time_per_roi=None):
+        """
+        Estimates the time for the rest of the acquisition.
+
+        :param site_idx: (int) index of the current site
+        :param actual_time_per_site: (float) the actual time spent for the current site
+        :return: (float) the estimated time for the rest of the acquisition
+        """
+        remaining_rois = (len(self.areas) - roi_idx)
+
+        if actual_time_per_roi:
+            acquisition_time = actual_time_per_roi * remaining_rois
+        else:
+            acquisition_time = estimate_autofocus_in_roi_time(self.n_focus_points, self._ccd) * remaining_rois
+
+        return acquisition_time
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+        """
+        self._future._task_state = RUNNING
+        da_rois = []
+        try:
+            actual_time_per_roi = None
+            # create a for loop for roi to create sub futures
+            for idx, roi in enumerate(self.areas):
+                start_time = time.time()
+                remaining_t = self.estimate_time(idx, actual_time_per_roi)
+                self._future.set_end_time(time.time() + remaining_t)
+
+                # cancel the sub future
+                if self._future._task_state == CANCELLED:
+                    raise CancelledError()
+
+                # run autofocus for the selected roi
+                self._future.running_subf = autofocus_in_roi(roi,
+                                                             self._stage,
+                                                             self._ccd,
+                                                             self._focus,
+                                                             self.focus_rng,
+                                                             self.n_focus_points,
+                                                             self.conf_level)
+
```

Comment:
Each check for cancel + start of a sub-future should be within a lock block (`with self._task_lock`).
This is the whole point of the lock: to avoid the possibility that someone cancels just before the start of the sub-future. With the lock, we always ensure that the task state is either checked before starting the sub-future, or the sub future is started, and it's immediately cancelled.